From 87c4b864ed9906d3e6bef4a5cc51931b808f9f78 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sat, 18 Jul 2026 23:24:23 -0400
Subject: [PATCH 202/285] winewayland: Only ensure contents for toplevel client
 surface.

child windows dont have their own window_surface.
---
 dlls/winewayland.drv/wayland_surface.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index 99afd91a4d2..e66bfc2c0f1 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -1426,11 +1426,11 @@ static const struct wl_buffer_listener dummy_buffer_listener =
 /**********************************************************************
  *          wayland_surface_ensure_contents
  *
  * Ensure that the wayland surface has up-to-date contents, by committing
  * a dummy buffer if necessary.
  */
-static BOOL wayland_surface_ensure_contents(struct wayland_surface *surface, BOOL fullscreen)
+static BOOL wayland_surface_ensure_contents(struct wayland_surface *surface, BOOL fullscreen, BOOL toplevel)
 {
     enum wl_shm_format format = fullscreen ? WL_SHM_FORMAT_XRGB8888 : WL_SHM_FORMAT_ARGB8888;
     enum wayland_surface_ensure_type needs_contents = WAYLAND_SURFACE_NOT_ENSURED;
     HWND hwnd = surface->hwnd;
@@ -1447,7 +1447,7 @@ static BOOL wayland_surface_ensure_contents(struct wayland_surface *surface, BOO
      * 1. fullscreen window black background
      * 2. avoid alpha blending between client surface and toplevel */
     if (surface->ensured_contents != WAYLAND_SURFACE_ENSURED_DUMMY_BUFFER &&
-        (fullscreen || EqualRect(&surface->window.client_rect, window)))
+        (fullscreen || (toplevel && EqualRect(&surface->window.client_rect, window))))
         needs_contents = WAYLAND_SURFACE_ENSURED_DUMMY_BUFFER;
 
     /* we still need a contents, so in the other cases we use
@@ -1504,7 +1504,7 @@ static void wayland_client_surface_present(struct client_surface *client, HDC hd
 
     if ((wayland_surface = data->wayland_surface))
     {
-        expose = wayland_surface_ensure_contents(wayland_surface, data->is_fullscreen);
+        expose = wayland_surface_ensure_contents(wayland_surface, data->is_fullscreen, hwnd == toplevel);
 
         /* Handle any processed configure request, to ensure the related
          * surface state is applied by the compositor. */
-- 
2.55.0

